kv/kvserver: TestTruncateLog failed

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4845647&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4845647&tab=artifacts#/) on release-22.1 @ [9b76da708050aab88c24eaa74e7cae05aa5ddc70](https://github.com/cockroachdb/cockroach/commits/9b76da708050aab88c24eaa74e7cae05aa5ddc70):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog3631979790
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:810: -- test log scope end --
--- FAIL: TestTruncateLog (0.63s)
=== RUN   TestTruncateLog/loosely-coupled=false
    raft_log_queue_test.go:771: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=false (0.29s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-14994

This failure is in loosely-coupled=false, unlike the TestTerm issue (which I was unable to reproduce and seems to not have recurred for 2 weeks).
```
    --- FAIL: TestTruncateLog/loosely-coupled=false (0.29s)
    --- PASS: TestTruncateLog/loosely-coupled=true (0.30s)
```
Trying to reproduce.
I see the following in the logs, so maybe the problem is in the test which got unlucky with an unexpected merge.
```
I220409 07:14:24.781245 630552 kv/kvserver/batcheval/cmd_truncate_log.go:56 ⋮ [s1,r1/1:‹/M{in-ax}›] 1  attempting to truncate raft logs for another range: r2. Normally this is due to a merge and can be ignored.
```
This is contradicting the place of the failure which is the error below, which suggest more has been truncated (not less).
```
		// The term of the last truncated entry is still available.
		tc.repl.mu.Lock()
		term, err := tc.repl.raftTermRLocked(indexes[4])
		tc.repl.mu.Unlock()
		if err != nil {
			t.Fatal(err)
		}
```

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=6484735&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=6484735&tab=artifacts#/) on release-22.1 @ [ae7edece60a9d0d89a86a6fc6408eb253c371881](https://github.com/cockroachdb/cockroach/commits/ae7edece60a9d0d89a86a6fc6408eb253c371881):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog4246184072
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:805: -- test log scope end --
--- FAIL: TestTruncateLog (0.73s)
=== RUN   TestTruncateLog/loosely-coupled=false
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=false (0.32s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=7018804&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=7018804&tab=artifacts#/) on release-22.1 @ [3875aa648b3aeb78aa458b3513e41bdbc52004e3](https://github.com/cockroachdb/cockroach/commits/3875aa648b3aeb78aa458b3513e41bdbc52004e3):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog3712744608
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:805: -- test log scope end --
--- FAIL: TestTruncateLog (0.67s)
=== RUN   TestTruncateLog/loosely-coupled=false
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=false (0.31s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9030843?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9030843?buildTab=artifacts#/) on release-22.1 @ [d21a4a188562e73172c66fa9d57eb1baa6ab149c](https://github.com/cockroachdb/cockroach/commits/d21a4a188562e73172c66fa9d57eb1baa6ab149c):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog684204320
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:805: -- test log scope end --
--- FAIL: TestTruncateLog (0.68s)
=== RUN   TestTruncateLog/loosely-coupled=true
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=true (0.33s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9355454?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9355454?buildTab=artifacts#/) on release-22.1 @ [0468eb91d6fa303a4f016e99aff7bb2824c38f1c](https://github.com/cockroachdb/cockroach/commits/0468eb91d6fa303a4f016e99aff7bb2824c38f1c):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog2078562329
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:805: -- test log scope end --
--- FAIL: TestTruncateLog (0.82s)
=== RUN   TestTruncateLog/loosely-coupled=false
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=false (0.36s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9419545?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9419545?buildTab=artifacts#/) on release-22.1 @ [21fc175f98866bd9b18ead1606867bdb933878e4](https://github.com/cockroachdb/cockroach/commits/21fc175f98866bd9b18ead1606867bdb933878e4):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog268465716
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:805: -- test log scope end --
--- FAIL: TestTruncateLog (0.82s)
=== RUN   TestTruncateLog/loosely-coupled=false
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=false (0.37s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9500984?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9500984?buildTab=artifacts#/) on release-22.1 @ [afad258bd2690d6ff8b0b8636224775dabbd6d98](https://github.com/cockroachdb/cockroach/commits/afad258bd2690d6ff8b0b8636224775dabbd6d98):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog2791901721
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:805: -- test log scope end --
--- FAIL: TestTruncateLog (0.80s)
=== RUN   TestTruncateLog/loosely-coupled=true
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=true (0.39s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestTruncateLog [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9556902?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9556902?buildTab=artifacts#/) on release-22.1 @ [9ab632f3bedfb653dacf571fb4b10e02de51ef77](https://github.com/cockroachdb/cockroach/commits/9ab632f3bedfb653dacf571fb4b10e02de51ef77):


```
=== RUN   TestTruncateLog
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTruncateLog2261802216
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTruncateLog
    raft_log_queue_test.go:805: -- test log scope end --
--- FAIL: TestTruncateLog (0.88s)
=== RUN   TestTruncateLog/loosely-coupled=false
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=false (0.38s)
=== RUN   TestTruncateLog/loosely-coupled=true
    raft_log_queue_test.go:766: requested index is unavailable due to compaction
    --- FAIL: TestTruncateLog/loosely-coupled=true (0.45s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTruncateLog.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/batcheval/cmd_truncate_log_test.go
+++ b/pkg/kv/kvserver/batcheval/cmd_truncate_log_test.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -55,6 +56,8 @@ func readTruncStates(
 func TestTruncateLog(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
+
+	skip.WithIssue(t, 79719)
 
 	ctx := context.Background()
 	const (

--- a/pkg/kv/kvserver/replica_test.go
+++ b/pkg/kv/kvserver/replica_test.go
@@ -859,15 +859,15 @@ func TestLeaseReplicaNotInDesc(t *testing.T) {
 }
 
 // TestReplicaRangeMismatchRedirect tests two behaviors that should occur.
-// - Following a Range split, the client may send BatchRequests based on stale
-//   cache data targeting the wrong range. Internally this triggers a
-//   RangeKeyMismatchError, but in the cases where the RHS of the range is still
-//   present on the local store, we opportunistically retry server-side by
-//   re-routing the request to the right range. No error is bubbled up to the
-//   client.
-// - This test also ensures that after a successful server-side retry attempt we
-//   bubble up the most up-to-date RangeInfos for the client to update its range
-//   cache.
+//   - Following a Range split, the client may send BatchRequests based on stale
+//     cache data targeting the wrong range. Internally this triggers a
+//     RangeKeyMismatchError, but in the cases where the RHS of the range is still
+//     present on the local store, we opportunistically retry server-side by
+//     re-routing the request to the right range. No error is bubbled up to the
+//     client.
+//   - This test also ensures that after a successful server-side retry attempt we
+//     bubble up the most up-to-date RangeInfos for the client to update its range
+//     cache.
 func TestReplicaRangeMismatchRedirect(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -7344,6 +7344,8 @@ func TestTerm(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
+	skip.WithIssue(t, 100449)
+
 	testutils.RunTrueAndFalse(t, "loosely-coupled", func(t *testing.T, looselyCoupled bool) {
 		ctx := context.Background()
 		tc := testContext{}
@@ -8696,13 +8698,12 @@ func TestRefreshFromBelowGCThreshold(t *testing.T) {
 // The test contains a subtest for each of the combinations of the following
 // boolean options:
 //
-// - followerRead: configures whether the read should be served from the
+//   - followerRead: configures whether the read should be served from the
 //     leaseholder replica or from a follower replica.
 //
-// - thresholdFirst: configures whether the GC operation should be split into
+//   - thresholdFirst: configures whether the GC operation should be split into
 //     two requests, with the first bumping the GC threshold and the second
 //     GCing the expired version. This is how the real MVCC GC queue works.
-//
 func TestGCThresholdRacesWithRead(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -13225,14 +13226,14 @@ func TestSplitSnapshotWarningStr(t *testing.T) {
 //
 // The test does the following things:
 //
-//  * Propose cmd at an initial MaxLeaseIndex.
-//  * Refresh that cmd immediately.
-//  * Fail the initial command with an injected error which will lead to a
-//    reproposal at a higher MaxLeaseIndex.
-//  * Simultaneously update the lease sequence number on the replica so all
-//    future commands will fail with NotLeaseHolderError.
-//  * Enable unconditional refreshes of commands after a raft ready so that
-//    higher MaxLeaseIndex commands are refreshed.
+//   - Propose cmd at an initial MaxLeaseIndex.
+//   - Refresh that cmd immediately.
+//   - Fail the initial command with an injected error which will lead to a
+//     reproposal at a higher MaxLeaseIndex.
+//   - Simultaneously update the lease sequence number on the replica so all
+//     future commands will fail with NotLeaseHolderError.
+//   - Enable unconditional refreshes of commands after a raft ready so that
+//     higher MaxLeaseIndex commands are refreshed.
 //
 // This order of events ensures that there will be a committed command which
 // experiences the lease mismatch error but does not carry the highest

